### PR TITLE
8270056: Generated lambda class can not access protected static method of target class

### PR DESCRIPTION
**Pull request by kelthuzadx**
Generated lambda class can not access protected static method of the target class. The following exception is thrown when executing the attached reproducible program:

```
Exception in thread "main" java.lang.IllegalAccessError: class AccessProtectedStaticMethodFromSuper$B$$Lambda$15/0x0000000800b8ea48 tried to access protected method 'void AccessProtectedStaticMethodFromSuper$A.func()' (AccessProtectedStaticMethodFromSuper$B$$Lambda$15/0x0000000800b8ea48 is in unnamed module of loader AccessProtectedStaticMethodFromSuper$1Loader @71dac704; AccessProtectedStaticMethodFromSuper$A is in unnamed module of loader AccessProtectedStaticMethodFromSuper$1Loader @39ed3c8d)
	at AccessProtectedStaticMethodFromSuper.main(AccessProtectedStaticMethodFromSuper.java:51)
```

This issue is similar to JDK-8254975(#767) with slight differences: generated lambda proxy calls static protected method rather than protected member method.

The proposed fix 1) tries to use MethodHandle instead of invoking forwardee directly(since the lambda class has no access to the resolved method) and 2) does not force accepting an implClass as the first argument when invoking a static method.

Testing:
- test/jdk/java/ with release mode
- presubmit tests

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8270056](https://bugs.openjdk.java.net/browse/JDK-8270056): Generated lambda class can not access protected static method of target class

### Reviewers
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**)

### Contributors
 * NekoCaffeine `<nekocaffeine@qq.com>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4714/head:pull/4714` \
`$ git checkout pull/4714`

Update a local copy of the PR: \
`$ git checkout pull/4714` \
`$ git pull https://git.openjdk.java.net/jdk pull/4714/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4714`

View PR using the GUI difftool: \
`$ git pr show -t 4714`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4714.diff">https://git.openjdk.java.net/jdk/pull/4714.diff</a>

</details>

### CODE REVIEW

**Review by mlchung — COMMENTED** [review]
Good catch.   It should check if the target class (not `implClass`) and the declaring class of the implementation's method handle are in the same package if it's protected.

I prefer to add the new test case in the existing test `test/jdk/java/lang/invoke/lambda/superProtectedMethod/SuperMethodTest.java`.   The test's name may be made clearer to rename to `ProtectedMethodInOtherPackage` (something like that)

File: test/jdk/java/lang/invoke/lambda/superProtectedMethod/ProtectedMethodInOtherPackage.java [hunk 1]
```diff
@@ -101,6 +101,30 @@ public static void splitPackage() throws Throwable {
         ((Runnable) get.invoke(b)).run();
     }
 
+    @Test
+    public static void splitPackage1() throws Throwable {
```
Comment (mlchung) [inline test/jdk/java/lang/invoke/lambda/superProtectedMethod/ProtectedMethodInOtherPackage.java:105]: perhaps the method name can be `protectedStaticMethodInSplitPackage`

**Review by mlchung — COMMENTED** [review]
The class name needs to be renamed as well.  Also the classname in the `@run` command.

**Review by mlchung — APPROVED** [review]
Looks good.

**Comment by kelthuzadx** [comment]
Thanks Mandy for review!

/integrate

**Pull request closed without merging**